keras/layers/core.py

```diff
@@ -60,7 +60,10 @@ def __init__(self, mask_value=0., **kwargs):
         self.mask_value = mask_value
 
     def compute_mask(self, inputs, mask=None):
-        return K.any(K.not_equal(inputs, self.mask_value), axis=-1)
+        output_mask = K.any(K.not_equal(inputs, self.mask_value), axis=-1)
+        if hasattr(inputs, '_keras_shape'):
+            output_mask._keras_shape = inputs._keras_shape[:-1]
+        return output_mask
```

Comment:
This change is specific to static shape inference in Theano. It should be made in `K.any` and `K.not_equal` (only in the Theano backend).